Framework/Beamline/src/DetectorInfo.cpp

```diff
@@ -78,28 +90,236 @@ bool DetectorInfo::isEquivalent(const DetectorInfo &other) const {
+Eigen::Quaterniond
+DetectorInfo::rotation(const std::pair<size_t, size_t> index) const {
+  return (*m_rotations)[linearIndex(index)];
+}
+
+/// Set the position of the detector with given index.
+void DetectorInfo::setPosition(const std::pair<size_t, size_t> index,
+                               const Eigen::Vector3d &position) {
+  m_positions.access()[linearIndex(index)] = position;
+}
+
+/// Set the rotation of the detector with given index.
+void DetectorInfo::setRotation(const std::pair<size_t, size_t> index,
+                               const Eigen::Quaterniond &rotation) {
+  m_rotations.access()[linearIndex(index)] = rotation.normalized();
+}
+
+/// Returns the scan count of the detector with given detector index.
+size_t DetectorInfo::scanCount(const size_t index) const {
+  if (!m_scanCounts)
+    return 1;
+  return (*m_scanCounts)[index];
+}
+
+/// Returns the scan interval of the detector with given index.
+std::pair<int64_t, int64_t>
+DetectorInfo::scanInterval(const std::pair<size_t, size_t> index) const {
+  if (!m_scanIntervals)
+    return {0, 0};
+  return (*m_scanIntervals)[linearIndex(index)];
+}
+
+/// Set the scan interval of the detector with given index.
+void DetectorInfo::setScanInterval(const std::pair<size_t, size_t> index,
+                                   std::pair<int64_t, int64_t> interval) {
+  if (!m_scanIntervals)
+    initScanIntervals();
+  // We forbid this since we (currently?) can not verify that the new interval
+  // has no collisions with any of the other intervals for this detector.
+  checkNoTimeDependence();
+  if (interval.first >= interval.second)
+    throw std::runtime_error(
+        "DetectorInfo: cannot set scan interval with start > end");
+  m_scanIntervals.access()[linearIndex(index)] = interval;
+}
+
+namespace {
+void failMerge(const std::string &what) {
+  throw std::runtime_error(std::string("Cannot merge DetectorInfo: ") + what);
+}
+
+std::pair<size_t, size_t>
+getIndex(const Kernel::cow_ptr<std::vector<std::pair<size_t, size_t>>> &indices,
+         const size_t index) {
+  if (!indices)
+    return {index, 0};
+  return (*indices)[index];
+}
+}
+
+/** Merges the contents of other into this.
+ *
+ * Scan intervals in both other and this must be set. Intervals must be
+ * identical or non-overlapping. If they are identical all other parameters (for
+ * that index) must match. */
+void DetectorInfo::merge(const DetectorInfo &other) {
+  if (size() != other.size())
+    failMerge("size mismatch");
+  if (!m_scanIntervals || !other.m_scanIntervals)
+    failMerge("scan intervals not defined");
+  if (!(m_isMonitor == other.m_isMonitor) &&
+      (*m_isMonitor != *other.m_isMonitor))
+    failMerge("monitor flags mismatch");
+  // TODO If we make masking time-independent we need to check masking here.
+
+  std::vector<bool> merge(other.m_positions->size(), true);
+
+  for (size_t index1 = 0; index1 < other.m_positions->size(); ++index1) {
```

Comment:
In the following there's some very confusing mixing of the `index` terminology. i.e. `index1` should be something like `linearIndex1`